protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,376 @@
+    // The score increases each time a remote peer successfully dials this address.
+    addresses: HashMap<Multiaddr, FiniteAddrScore>,
+    // Ongoing inbound requests, where no response has been sent back to the remote yet.
+    ongoing_inbound: HashMap<PeerId, ResponseChannel<DialResponse>>,
+    // Ongoing outbound dial-requests, where no response has been received from the remote yet.
+    ongoing_outbound: HashSet<PeerId>,
+    // Recently connected peers to which we want to send a dial-request.
+    pending_requests: VecDeque<PeerId>,
+}
+
+impl Default for Behaviour {
+    fn default() -> Self {
+        Behaviour::new(Config::default())
+    }
+}
+
+impl Behaviour {
+    pub fn new(config: Config) -> Self {
+        let protocols = iter::once((AutoNatProtocol, ProtocolSupport::Full));
+        let mut cfg = RequestResponseConfig::default();
+        cfg.set_request_timeout(config.timeout);
+        let inner = RequestResponse::new(AutoNatCodec, protocols, cfg);
+        Self {
+            inner,
+            addresses: HashMap::default(),
+            ongoing_inbound: HashMap::default(),
+            ongoing_outbound: HashSet::default(),
+            pending_requests: VecDeque::default(),
+        }
+    }
+
+    /// Add a new address to the address list that is send to remote peers in a dial-request.
+    pub fn add_local_address(&mut self, address: Multiaddr) {
+        if self.addresses.get(&address).is_none() {
+            self.addresses.insert(address, 1);
+        }
+    }
+
+    /// Get the list of local addresses with their current score.
+    ///
+    /// The score of an address increases each time a remote peer successfully dialed us via this address.
+    /// Therefore higher scores indicate a higher reachability.
+    pub fn address_list(&self) -> impl Iterator<Item = (&Multiaddr, &FiniteAddrScore)> {
+        self.addresses.iter()
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ProtocolsHandler = <RequestResponse<AutoNatCodec> as NetworkBehaviour>::ProtocolsHandler;
+    type OutEvent = ();
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        self.inner.new_handler()
+    }
+
+    fn addresses_of_peer(&mut self, peer: &PeerId) -> Vec<Multiaddr> {
+        self.inner.addresses_of_peer(peer)
+    }
+
+    fn inject_connected(&mut self, peer: &PeerId) {
+        self.inner.inject_connected(peer)
+    }
+
+    fn inject_disconnected(&mut self, peer: &PeerId) {
+        self.inner.inject_disconnected(peer);
+        self.ongoing_inbound.remove(peer);
+    }
+
+    fn inject_connection_established(
+        &mut self,
+        peer: &PeerId,
+        conn: &ConnectionId,
+        endpoint: &ConnectedPoint,
+    ) {
+        self.inner
+            .inject_connection_established(peer, conn, endpoint);
+
+        // Initiate a new dial request if there is none pending.
+        if !self.ongoing_outbound.contains(peer) {
+            self.pending_requests.push_back(*peer);
+        }
```

Comment:
Just as a note, though this likely requires a lot more design work: We will need AutoNAT to do continuous probing, among other things, to keep mappings on NAT devices active.